### PR TITLE
Introducing initial tentative stub for #463: psalm-specific type declarations

### LINKED ISSUE #463: Provide `@psalm-param class-string<T>` documentation for proxy factories and anything type-hinted `object` and `: object`
See https://psalm.dev/docs/templated_annotations/#param-class-stringlttgt

### PR DESCRIPTION
**Pull request by Ocramius**
Fixes #463

This is an incomplete patch that just serves to display that types can be
improved massively in here, making even this magic contraption of a library
usable downstream with some decent type constraints.

TODOs:

 - [x] need to implement CI steps to ensure type inference works as expected
 - [x] full type coverage in ProxyManager internals (note: no tests yet - those require more work)
 - [x] pin to stable `vimeo/psalm` (`^3.4.0`)
 - [x] split type check test file into multiple type-checked examples
 - [x] raise psalm reporting levels
 - [x] add utilities for PHPStorm type inference

Blocked by:

 - [x] https://github.com/vimeo/psalm/issues/1673
 - [x] https://github.com/vimeo/psalm/issues/1674
 - [x] https://github.com/vimeo/psalm/issues/1675
 - [x] https://github.com/vimeo/psalm/issues/1677
 - [x] https://github.com/vimeo/psalm/issues/1679
 - [x] https://github.com/vimeo/psalm/issues/1680
 - [x] https://github.com/vimeo/psalm/issues/1681
 - [x] https://github.com/vimeo/psalm/issues/1682
 - [x] https://github.com/vimeo/psalm/issues/1683
 - [x] https://github.com/vimeo/psalm/issues/1722
 - [x] https://github.com/vimeo/psalm/pull/1726

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
         "phpstan/phpstan":              "^0.11.4",
         "phpstan/phpstan-phpunit":      "^0.11",
         "infection/infection":          "^0.12.2",
-        "symfony/console":              "^4.2.4"
+        "symfony/console":              "^4.2.4",
+        "vimeo/psalm": "^3.3"
```
Comment (Ocramius) [inline composer.json]: Note: I just committed whatever I had stashed locally, so I'd just need a rough high-level review of this feature.

File: type-example.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        $initializer   = null; // disable initialization
+        $wrappedObject = new MyProxiedClass();
+
+        return true;
+    })
+    ->sayHello();
+
+echo (new NullObjectFactory())
+    ->createProxy(MyProxiedClass::class)
+    ->sayHello();
```
Comment (Ocramius) [inline type-example.php]: This type check seems to fail: `NullObjectFactory#createProxy()` may be annotated incorrectly.

File: type-example.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        $initializer = function (
+            GhostObjectInterface $proxy,
+            string $method,
+            array $parameters,
+            & $initializer,
+            array $properties
+        ) {
+            $initializer = null; // disable initialization
+        })
+    ->sayHello();
```
Comment (Ocramius) [inline type-example.php]: This type declaration seems to fail type checks:

```
ERROR: UndefinedInterfaceMethod - type-example.php:49:7 - Method ProxyManager\Proxy\GhostObjectInterface::sayhello does not exist
    ->sayHello();
```

@muglug I'm sorry to chase you across repositories, but do you maybe know whether this is a docblock parser bug? This is the docblock:

https://github.com/Ocramius/ProxyManager/blob/300973221ee90f0f126e5f3b2753410c7c5c88d4/src/ProxyManager/Factory/LazyLoadingGhostFactory.php#L38-L88

Comment (muglug) [reply type-example.php]: that docblock should be
```php
/**
 * @psalm-template RealObjectType 
 * 
 * @psalm-param class-string<RealObjectType> $className 
 * @psalm-param array{skippedProperties?: array<int, string>} $proxyOptions
 * 
 * @psalm-return RealObjectType&GhostObjectInterface 
 */ 
```

Comment (Ocramius) [reply type-example.php]: Daaang, so it is just that `?`!

Comment (muglug) [reply type-example.php]: And missing `$proxyOptions`

**Comment by stof** [comment]
a PR fixing itself ? That's inception.

**Comment by Ocramius** [comment]
> a PR fixing itself ?

What do you mean by that? I didn't catch the joke XD

File: src/ProxyManager/Inflector/ClassNameInflector.php [hunk 4]
```diff
@@ -37,14 +38,14 @@ public function getUserClassName(string $className) : string
         $className = ltrim($className, '\\');
         $position  = strrpos($className, $this->proxyMarker);
 
-        if ($position === false) {
+        if (! is_int($position)) {
```
Comment (stof) [inline src/ProxyManager/Inflector/ClassNameInflector.php]: is this change actually needed by psalm ? It should be able to detect that the other case has an int when comparing to false, as the signature is `int|false`.

**Comment by stof** [comment]
The description contains this:

> Fixes #465

This reference is the PR itself.

**Comment by Ocramius** [comment]
Ha, fixed, thanks!

File: src/ProxyManager/Inflector/ClassNameInflector.php (hunk 4 above)
Comment (Ocramius) [reply src/ProxyManager/Inflector/ClassNameInflector.php]: The inference for now is `int|bool`, not `int|false`. I'm fine with adapting the code for these scenarios for now, as upstream work for PHPStorm stubs is ongoing.

Comment (stof) [reply src/ProxyManager/Inflector/ClassNameInflector.php]: oh, so PHPStan has the `int|false` type for `strrpos`, but psalm does not have it yet ?

Comment (muglug) [reply src/ProxyManager/Inflector/ClassNameInflector.php]: Psalm definitely has it - @Ocramius is maybe talking about PHPStorm?

Comment (Ocramius) [reply src/ProxyManager/Inflector/ClassNameInflector.php]: Well, this has been inferred as `int|bool` here :|

Comment (muglug) [reply src/ProxyManager/Inflector/ClassNameInflector.php]: Maybe something else is happening? https://psalm.dev/r/d029859bdf works as expected

Comment (muglug) [reply src/ProxyManager/Inflector/ClassNameInflector.php]: Or possibly this was an  attempt to fix the possibly-false operandi issue (that you fixed with an explicit `(int)` cast below

**Comment by Ocramius** [comment]
```
Checks took 0.66 seconds and used 153.505MB of memory
Psalm was able to infer types for 99.7795% of the codebase
```

This is :fire:

File: psalm.xml [hunk 5]
```diff
@@ -6,55 +6,63 @@
...
 
-        <RedundantCondition errorLevel="suppress" />
+        <PropertyNotSetInConstructor errorLevel="suppress"/>
+        <MissingConstructor errorLevel="error"/>
+        <MissingClosureParamType errorLevel="info"/>
+        <MissingParamType errorLevel="info"/>
 
-        <DocblockTypeContradiction errorLevel="error" />
-        <RedundantConditionGivenDocblockType errorLevel="error" />
+        <RedundantCondition errorLevel="suppress"/>
```
Comment (muglug) [inline psalm.xml]: Is this emitted somewhere erroneously? It's the only issue here that should never be emitted for good code (so if it is, that's a Psalm bug)

**Comment by Ocramius** [comment]
Ok, so my latest changes lead to wild issues in Psalm:

```
ERROR: UndefinedMethod - tests/static-analysis/lazy-loading-value-holder.php:57:25 - Method ProxyManager\Example\VirtualProxy\Foo::sayhello does not exist
    echo $wrappedValue->sayHello();
```

There is no mention of `ProxyManager\Example\VirtualProxy\Foo` in that file, to this feels like a global multi-file shared state issue. Not sure how to reproduce it for psalm to consume.

File: psalm.xml (hunk 5 above)
Comment (Ocramius) [reply psalm.xml]: Will need to inspect this: I plan to raise the levels before merging

**Comment by muglug** [comment]
> this feels like a global multi-file shared state issue

Yeah something in examples is polluting stuff - I'll find and fix

**Comment by muglug** [comment]
Fixed in https://github.com/vimeo/psalm/commit/0ad57695a893aa1a43a791777ce621aa493164a1

File: psalm.xml (hunk 5 above)
Comment (Ocramius) [reply psalm.xml]: Bumped all inspection levels

**Comment by Ocramius** [comment]
Hmm, since latest builds, psalm now fails with something like this on 7.4:

```
$ ./vendor/bin/psalm
PHP Fatal error:  Uncaught ParseError: syntax error, unexpected 'Fn' (T_FN), expecting identifier (T_STRING) in /home/travis/build/Ocramius/ProxyManager/vendor/vimeo/psalm/src/Psalm/Internal/Provider/ReturnTypeProvider/ArrayFilterReturnTypeProvider.php:64
```

@muglug I just need to pin this to a stable release with a green build, and then I can release here :+1:

**Comment by Ocramius** [comment]
I'll send a patch to fix the `Fn` issue btw :+1:

**Comment by dkarlovi** [comment]
> a PR fixing itself ? That's inception.

@Ocramius is just that good.

**Comment by Ocramius** [comment]
Green! Now needs a release upstream and we can :ship: \o/

File: src/ProxyManager/Inflector/ClassNameInflector.php (hunk 4 above)
Comment (Ocramius) [reply src/ProxyManager/Inflector/ClassNameInflector.php]: Keeping this as-is for now.

**Comment by muglug** [comment]
Will hopefully release later today, barring any major impediments

**Comment by muglug** [comment]
Boom: https://github.com/vimeo/psalm/releases/tag/3.4.0

**Comment by Ocramius** [comment]
Thanks! That should be enough to do a release here, if all is green :rocket:

**Comment by Ocramius** [comment]
Relevant CI tasks are GO! :+1:
